api/src/OpenTelemetry/Trace/Core.hs

```diff
@@ -375,9 +388,11 @@ inSpan'' t cs n args f = do
         pure (lookupSpan ctx, s)
     )
     ( \e (parent, s) -> liftIO $ do
-        forM_ e $ \(SomeException inner) -> do
+        forM_ e $ \anException@(SomeException inner) -> do
           setStatus s $ Error $ T.pack $ displayException inner
-          recordException s [] Nothing inner
+          let exceptionAttributes = concatMap (\enricher -> catchesHandler enricher anException) $
+                V.toList (tracerProviderExceptionEnrichers $ tracerProvider t)
+          recordException s exceptionAttributes Nothing inner
         endSpan s Nothing
         adjustContext $ \ctx ->
           maybe (removeSpan ctx) (`insertSpan` ctx) parent
```

Comment:
I feel like what I *usually* see is a configurable function, like `data Settings = Settings { settingsOnException :: SomeException -> IO () }`. 

For this pattern, we can do:

```haskell
data TracerProvider = TracerProvider 
  { ...
  , tracerProviderExceptionToAttributes :: SomeException -> [(Text, Attr)]
  }
```

Which removes the need for the `ExceptionEnricher` complexity.

But, is this equivalent?

Let's write one that pulls annotations off an annotated exception.

```haskell
tracerProvider = TracerProvicer {..}
  where
    tracerProviderExceptionToAttributes somE@(SomeException inner) 
      -- because fromException always succeeds, don't want to infinite recursion
      | Just (AnnotatedException anns err) <- cast inner 
      = convertAnnotations anns <> tracerProvicerExceptionToAttributes (toException err)
      | Just (SomeAsyncException inner) <- fromException somE
      = ("Async Exception", _) : tracerProviderExceptionToAttributes (SomeException inner)
      -- use `SomeException` instead of `toException` because `toException` will rewrap in `SomeAsyncException`
```

Hmm, already getting pretty complex...

But I think it's probably better than the `ExceptionEnricher` class, which doesn't locally know what the other things are doing.

```haskell
instance HasExceptionEnricher SomeAsyncException where
    exceptionEnricher _ = ExceptionEnricher 
      do \(SomeAsyncException e) -> 
           -- well, we can't really do anything interesting here
      do "SomeAsyncException"
```

`AnnotatedException` *is* able to summon the canonical `ExceptionEnricher` for a type, since it has the type parameter. But no other exception wrapper would work.

Reply:
I suppose we could work around the wrapper issue by passing in known exception enrichers to the function:

``` haskell
data ExceptionEnricher = forall e. Exception e => ExceptionEnricher
  { exceptionEnricherAdditionalFields :: [ExceptionEnricher] {- ^ all known enrichers -} -> e -> [(Text, Attribute)]
  , exceptionEnricherName :: Text
  }
```

That would allow us in the case of wrappers to feed the inner exception through the same process.